crates/kubelet/src/pod/status.rs

```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
     reason: &str,
     container_statuses: Vec<KubeContainerStatus>,
     init_container_statuses: Vec<KubeContainerStatus>,
-) -> serde_json::Value {
-    serde_json::json!(
-       {
-           "metadata": {
-               "resourceVersion": "",
-           },
-           "status": {
-               "phase": phase,
-               "reason": reason,
-               "containerStatuses": container_statuses,
-               "initContainerStatuses": init_container_statuses,
-           }
-       }
-    )
+) -> Status {
+    let mut status = Status::new();
+    status.set_phase(phase);
+    status.set_reason(reason);
+    status.set_container_statuses(container_statuses);
+    status.set_init_container_statuses(init_container_statuses);
+    status
 }
 
-/// Describe the status of a workload.
 #[derive(Clone, Debug, Default)]
-pub struct Status {
-    /// Allows a provider to set a custom message, otherwise, kubelet will infer
-    /// a message from the container statuses
-    pub message: StatusMessage,
-    /// The statuses of containers keyed off their names
-    pub container_statuses: ContainerMap<ContainerStatus>,
-}
+/// Pod Status wrapper.
+pub struct Status(KubePodStatus);
 
-#[derive(Clone, Debug)]
-/// The message to be set in a pod status update.
-pub enum StatusMessage {
-    /// Do not change the existing status message.
-    LeaveUnchanged,
-    /// Remove any existing status message.
-    Clear,
-    /// Set the status message to the given value.
-    Message(String),
-}
+impl Status {
+    /// Create a new status with no fields set.
+    pub fn new() -> Self {
+        Status(Default::default())
+    }
 
-impl Default for StatusMessage {
-    fn default() -> Self {
-        Self::LeaveUnchanged
+    /// Set Pod phase.
+    pub fn set_phase(&mut self, phase: Phase) {
```

Comment:
I think we could basically rename this struct to `StatusBuilder` then. I can if you want. The reason I did this is that there is a distinction between a field that was never set, or is explicitly being set to "", this results in patches overwriting things they shouldnt. There might be a simpler way to do this, and I didnt do the same thing for containers because its much less important. Thoughts?

Reply:
I personally think this:

```rust
let status = StatusBuilder::new().phase(...).message(...).build();
```

reads much better (and is equivalent to):

```rust
let mut status = Status::new();
status.set_phase(...);
status.set_message(...);
```

I figure the point about an empty field is pretty easy to handle. You can still use the setter/builder to explicitly set an empty string and leave it empty otherwise. However, this is completely a bikeshedding, so feel no pressure to change it

Reply:
Yep, I will refactor it to a Builder, I was just explaining why it got so complicated.